docs/architecture/ADR-008-borrow-tracking.md

```diff
@@ -0,0 +1,117 @@
+
+This change in borrow storage will be opaque to other parts of the leverage module: functions like `GetBorrow`, `GetTotalBorrows`, and `SetBorrow` will use `InterestScalar` internally, such that their inputs and outputs contain the real borrowed amount.
+
+As a result of efficiency gains, the parameter `InterestEpoch` will be removed, with periodic functions taking place every block.
+
+Additionally, the values `BorrowAPY`, `LendAPY`, and `uTokenExchangeRate` will be removed from state, instead being efficiently calculated when needed.
+
+## Detailed Design
+
+This decision mainly updates existing features, rather than adding new ones. The following changes are required to update the leverage module:
+
+**Params:**
+
+- Remove `InterestEpoch`
+
+**State:**
+
+- Add `InterestScalar` and `TotalAdjustedBorrows` to state, and add `Get/Set` functions
+- Remove `BorrowAPY`, `LendAPY`, and `uTokenExchangeRate` from state, and remove `Set` functions
+
+**Getters:**
+
+- Modify `Get` functions for `BorrowAPY`, `LendAPY`, and `uTokenExchangeRate` to calculate values in real time.
+- Modify `GetBorrow(addr,denom)`, `GetBorrowerBorrows(addr)`, and `GetAllBorrows()` to use `InterestScalar` and `AdjustedBorrow`
+- Modify `GetTotalBorrows(denom)` to use `InterestScalar` and `TotalAdjustedBorrows`
+
+**Setters:**
+
+- Modify `SetBorrow(addr,denom)` to use `InterestScalar` and a stored `AdjustedBorrow`, and to additionally update `TotalAdjustedBorrows` based on the difference from the previous value (calls get before set.)
+
+**Genesis:**
+
+- Rename the `Borrow` struct in genesis state to `AdjustedBorrow`, with `Amount` field changing to `sdk.Dec` from `sdk.Int`
+
+**Invariants:**
+
+- Add invariant which checks `TotalAdjustedBorrows` against the sum of all `AdjustedBorrow` returned by `GetAllBorrows()`
+
+### Example Scenarios
+
+The following example scenario should help clarify the meaning of the `AdjustedBorrow` and `InterestScalar`:
+
+> Assume a fresh system, containing a token denom `ured` which has never been borrowed or accrued interest before.
+>
+> By definition, `TotalAdjustedBorrowed("ured") = 0.0` and `InterestScalar("ured") = 1.0`
+>
+> User `Alice` borrows `1000 ured`. This information is stored in state as `AdjustedBorrow(alice,"ured") = 1000.000`, because adjusted borrow amount is real borrow amount divided by interest scalar. As a result, `TotalAdjustedBorrow("ured") = 1000.000`. Interest scalar is unchanged by borrowing.
+>
+> User `Bob` also borrows `2000 ured`, which is stored as  `AdjustedBorrow(alice,"ured") = 2000.000`. As a result, `TotalAdjustedBorrow("ured") = 3000.000`. Interest scalar is unchanged.
+>
+> Suppose that the interest rate for `ured` borrows works out to be `0.0003% per block`. On the next EndBlock, `InterestScalar("ured") *= 1.000003`. Both `AdjustedBorrow` values and `TotalAdjustedBorrow` are unchanged.
```

Comment:
```suggestion
> Assume a fresh system, containing a token denom `ured` which has never been borrowed or accrued interest before.
>
> By definition, `TotalAdjustedBorrowed("ured") = 0.0` and `InterestScalar("ured") = 1.0`
>
> User `Alice` borrows `1000 ured`. This information is stored in state as `AdjustedBorrow(alice,"ured") = 1000.000`, because adjusted borrow amount is real borrow amount divided by interest scalar. As a result, `TotalAdjustedBorrow("ured") = 1000.000`. Interest scalar is unchanged by borrowing.
>
> User `Bob` also borrows `2000 ured`, which is stored as  `AdjustedBorrow(bob,"ured") = 2000.000`. As a result, `TotalAdjustedBorrow("ured") = 3000.000`. Interest scalar is unchanged.
>
> Suppose that the interest rate for `ured` borrows works out to be `0.0003% per block`. On the next EndBlock, `InterestScalar("ured") *= 1.000003`. Both `AdjustedBorrow` values and `TotalAdjustedBorrow` are unchanged.
```